src/AMAMM.sol

```diff
@@ -2,69 +2,191 @@
+
+    using SafeCastLib for *;
+    using FixedPointMathLib for *;
+
+    /// -----------------------------------------------------------------------
+    /// Constants
+    /// -----------------------------------------------------------------------
+
+    function K(PoolId) internal view virtual returns (uint40) {
+        return 24;
+    }
+
+    function EPOCH_SIZE(PoolId) internal view virtual returns (uint256) {
+        return 1 hours;
+    }
+
+    function MIN_BID_MULTIPLIER(PoolId) internal view virtual returns (uint256) {
+        return 1.1e18;
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Storage variables
+    /// -----------------------------------------------------------------------
+
+    mapping(PoolId id => uint40) internal _lastUpdatedEpoch;
+    mapping(address deposits => uint256) public _userBalance;
+    mapping(Currency currency => uint256) internal _totalFees;
+    mapping(PoolId id => mapping(uint40 => Bid)) public poolEpochManager;
+    mapping(address manager => mapping(Currency currency => uint256)) internal _fees;
+
+    /// -----------------------------------------------------------------------
+    /// Getter actions
+    /// -----------------------------------------------------------------------
+
+    function getManager(PoolId id, uint40 epoch) public view returns (Bid memory) {
+        return poolEpochManager[id][epoch];
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Bidder actions
+    /// -----------------------------------------------------------------------
+    /// @inheritdoc IAmAmm
+    function bid(PoolId id, bytes7 payload, uint128 rent, uint128 deposit, uint40 _epoch) external virtual override {
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        _userBalance[msgSender] += _getDeposit(id, _epoch);
+
+        // ensure bid is valid
+        // - manager can't be zero address
+        // - bid needs to be greater than the next bid by >10%
+        // - deposit needs to cover the rent for K hours
+        // - deposit needs to be a multiple of rent
+        // - payload needs to be valid
+        if (
+            rent <= poolEpochManager[id][_epoch].rent.mulWad(MIN_BID_MULTIPLIER(id)) || deposit < rent * K(id)
+                || deposit % rent != 0 || !_payloadIsValid(id, payload)
+        ) {
+            revert AmAmm__InvalidBid();
+        }
+
+        uint256 prevWinner = _getDeposit(id, _epoch);
+
+        if (prevWinner == 0 && _getEpoch(id, block.timestamp) - _lastUpdatedEpoch[id] < K(id)) {
+            if (poolEpochManager[id][_lastUpdatedEpoch[id]].rent < rent && _findUpperManager(id, _epoch) == 0) {
```

Comment:
@leckylao I added `_findUpperManager` for the case that we are sandwiching 2 bids. COnsider the scenario: 
K is 10 and there is a manager at 0-10 epochs and another manager at 12-22 epochs. If I want to add a new bid at 5-15 then I would be interfering in 2 different managers. I can tell if there is a manager before me using `lastupdatedepoch` but there is no way to tell if there is a bid at the upper end. 
lmk if you think there is a better way to handle this

Reply:
Yeah, that would be tricky; for MVP and simplicity, let's restrict that you could only bid for epoch that's greater than `lastUpdatedEpoch`, then it would not have this scenario.

Reply:
I just think of a better way. Instead of the above restriction, we could limit that user could only bit lastUpdatedEpoch/currentEpoch + k whichever the highest. This way it would also restrict the conflict scenario.